File: src/main/java/io/github/slimefunguguproject/bump/utils/tags/TagParser.java
package io.github.slimefunguguproject.bump.utils.tags;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import java.util.function.BiConsumer;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.stream.JsonReader;

import org.bukkit.Keyed;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Tag;

import io.github.slimefunguguproject.bump.api.exceptions.TagMisconfigurationException;
import io.github.slimefunguguproject.bump.implementation.Bump;
import io.github.slimefunguguproject.bump.utils.JsonUtils;
import io.github.slimefunguguproject.bump.utils.Patterns;
import io.github.thebusybiscuit.slimefun4.libraries.dough.common.CommonPatterns;

/**
 * The {@link TagParser} is responsible for parsing a JSON file into a {@link BumpTag}.
 *
 * @author ybw0014
 * @see BumpTag
 */
class TagParser implements Keyed {
    /**
     * Every {@link Tag} has a {@link NamespacedKey}.
     * This is the {@link NamespacedKey} for the resulting {@link Tag}.
     */
    private final NamespacedKey key;

    /**
     * This constructs a new {@link TagParser} for the given {@link BumpTag}
     *
     * @param tag The {@link BumpTag} to parse inputs for.
     */
    TagParser(@Nonnull BumpTag tag) {
        this.key = tag.getKey();
    }

    /**
     * This will parse the JSON and run the provided callback with {@link Set Sets} of
     * matched {@link Material Materials} and {@link Tag Tags}.
     *
     * @param callback A callback to run after successfully parsing the input.
     * @throws TagMisconfigurationException This is thrown whenever the given input is malformed or no adequate
     *                                      {@link Material} or {@link Tag} could be found
     */
    public void parse(@Nonnull BiConsumer<Set<Material>, Set<Tag<Material>>> callback) throws TagMisconfigurationException {
        String path = "/tags/" + key.getKey() + ".json";
        Set<Material> materials = EnumSet.noneOf(Material.class);
        Set<Tag<Material>> tags = new HashSet<>();

        try {
            final InputStream stream = Bump.class.getResourceAsStream(path);
            JsonReader reader = new JsonReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
            JsonObject root = JsonUtils.parseReader(reader).getAsJsonObject();
            JsonElement child = root.get("values");

            if (child instanceof JsonArray) {
                JsonArray values = child.getAsJsonArray();

                for (JsonElement element : values) {
                    if (element instanceof JsonPrimitive primitive && primitive.isString()) {
                        parseString(element.getAsString(), materials, tags);
                    } else {
                        throw new TagMisconfigurationException(key, "Unexpected value format: " + element.getClass().getSimpleName() + " - " + element);
                    }
                }

                callback.accept(materials, tags);
            } else {
                throw new TagMisconfigurationException(key, "No values array specified");
            }
        } catch (NullPointerException | IllegalStateException | JsonParseException x) {
            throw new TagMisconfigurationException(key, x);
        }
    }

    @ParametersAreNonnullByDefault
    private void parseString(String value, Set<Material> materials, Set<Tag<Material>> tags) throws TagMisconfigurationException {
        if (Patterns.MINECRAFT_NAMESPACEDKEY.matcher(value).matches()) {
            Material material = Material.matchMaterial(value);

            if (material != null) {
                materials.add(material);
            }
        } else if (Patterns.BUMP_TAG.matcher(value).matches()) {
            String keyValue = CommonPatterns.COLON.split(value)[1].toUpperCase(Locale.ROOT);
            BumpTag tag = BumpTag.getTag(keyValue);

            if (tag != null) {
                tags.add(tag);
            } else {
                throw new TagMisconfigurationException(key, "Bump tag not exist: " + keyValue);
            }
        }
    }

    @Nonnull
    @Override
    public NamespacedKey getKey() {
        return key;
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/utils/tags/TagParser.kt
package io.github.slimefunguguproject.bump.utils.tags

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import com.google.gson.stream.JsonReader
import io.github.slimefunguguproject.bump.Bump
import io.github.slimefunguguproject.bump.api.exceptions.TagMisconfigurationException
import io.github.slimefunguguproject.bump.utils.JsonUtils
import io.github.slimefunguguproject.bump.utils.constant.Patterns
import io.github.thebusybiscuit.slimefun4.libraries.dough.common.CommonPatterns
import org.bukkit.Keyed
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Tag
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.function.BiConsumer

/**
 * The [TagParser] is responsible for parsing a JSON file into a [BumpTag].
 *
 * @author ybw0014
 * @see BumpTag
 */
internal class TagParser(tag: BumpTag) : Keyed {
    /**
     * Every [Tag] has a [NamespacedKey].
     * This is the [NamespacedKey] for the resulting [Tag].
     */
    private val tagKey = tag.key

    /**
     * This will parse the JSON and run the provided callback with [Sets][Set] of
     * matched [Materials][Material] and [Tags][Tag].
     *
     * @param callback A callback to run after successfully parsing the input.
     *
     * @throws TagMisconfigurationException This is thrown whenever the given input is malformed or no adequate
     * [Material] or [Tag] could be found
     */
    fun parse(callback: BiConsumer<Set<Material>, Set<Tag<Material>>>) {
        val path = "/tags/" + key.key + ".json"
        val materials: MutableSet<Material> = mutableSetOf()
        val tags: MutableSet<Tag<Material>> = mutableSetOf()

        try {
            val stream = Bump::class.java.getResourceAsStream(path) ?: error("Tag file not found within jar: $path")
            val reader = JsonReader(InputStreamReader(stream, StandardCharsets.UTF_8))
            val root: JsonObject = JsonUtils.parseReader(reader).asJsonObject
            val child = root["values"]

            if (child !is JsonArray) {
                error("No values array specified")
            }

            val values = child.getAsJsonArray()

            for (element in values) {
                if (element is JsonPrimitive && element.isString) {
                    parseString(element.getAsString(), materials, tags)
                } else {
                    error("Unexpected value format: " + element.javaClass.simpleName + " - " + element)
                }
            }

            callback.accept(materials, tags)
        } catch (ex: Exception) {
            throw TagMisconfigurationException(key, ex)
        }
    }

    private fun parseString(value: String, materials: MutableSet<Material>, tags: MutableSet<Tag<Material>>) {
        if (Patterns.MINECRAFT_NAMESPACEDKEY.matcher(value).matches()) {
            val material = Material.matchMaterial(value)

            if (material != null) {
                materials.add(material)
            }
        } else if (Patterns.BUMP_TAG.matcher(value).matches()) {
            val keyValue = CommonPatterns.COLON.split(value)[1].uppercase()
            val tag = BumpTag.getTag(keyValue)

            if (tag != null) {
                tags.add(tag)
            } else {
                error("Bump tag not exist: $keyValue")
            }
        }
    }

    override fun getKey() = tagKey
}
